ovara-ui/src/app/lib/http-client.ts

```diff
@@ -0,0 +1,172 @@
+    window.location.replace(loginUrl.toString());
+  }
+};
+
+const makeBareRequest = (request: Request) => {
+  request.headers.set('Caller-Id', '1.2.246.562.10.00000000001.ovara-ui');
+  if (['POST', 'PUT', 'PATCH', 'DELETE'].includes(request.method)) {
+    const csrfCookie = getCookies()['CSRF'];
+    if (csrfCookie) {
+      request.headers.set('CSRF', csrfCookie);
+    }
+  }
+
+  return doFetch(request);
+};
+
+const retryWithLogin = async (request: Request, loginUrl: string) => {
+  await makeBareRequest(new Request(loginUrl));
+  return makeBareRequest(request);
+};
+
+type BodyParser<T> = (res: Response) => Promise<T>;
+
+const TEXT_PARSER = async (res: Response) => await res.text();
+const BLOB_PARSER = async (response: Response) => await response.blob();
+
+const RESPONSE_BODY_PARSERS: Record<string, BodyParser<unknown>> = {
+  'application/json': async (response: Response) => await response.json(),
+  'application/octet-stream': BLOB_PARSER,
+  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet':
+    BLOB_PARSER,
+  'text/plain': TEXT_PARSER,
+};
+
+const responseToData = async <Result = unknown>(res: Response) => {
+  // Oletetaan JSON-vastaus, jos content-type header puuttuu
+  const contentType =
+    res.headers.get('content-type')?.split(';')?.[0] ?? 'text/plain';
+
+  const parseBody = (RESPONSE_BODY_PARSERS?.[contentType] ??
+    TEXT_PARSER) as BodyParser<Result>;
+
+  try {
+    return {
+      headers: res.headers,
+      data: await parseBody(res),
+    };
+  } catch (e) {
+    console.error(`Parsing fetch response body as "${contentType}" failed!`);
+    return Promise.reject(e);
+  }
+};
+
+const LOGIN_MAP = [
+  { urlIncludes: '/ovara-backend', loginUrl: configuration.ovaraBackendLogin },
+] as const;
+
+const makeRequest = async <Result>(request: Request) => {
+  const originalRequest = request.clone();
+  try {
+    const response = await makeBareRequest(request);
+    const responseUrl = new URL(response.url);
+    if (
+      isRedirected(response) &&
+      responseUrl.pathname.startsWith('/cas/login')
+    ) {
+      redirectToLogin();
+    }
+    return responseToData<Result>(response);
+  } catch (error: unknown) {
+    if (error instanceof FetchError) {
+      if (isUnauthenticated(error.response)) {
+        try {
+          for (const { urlIncludes, loginUrl } of LOGIN_MAP) {
+            if (request?.url?.includes(urlIncludes)) {
+              const resp = await retryWithLogin(request, loginUrl);
+              return responseToData<Result>(resp);
+            }
+          }
+        } catch (e) {
+          if (e instanceof FetchError && isUnauthenticated(e.response)) {
+            redirectToLogin();
+          }
+          return Promise.reject(e);
+        }
+      } else if (
+        isRedirected(error.response) &&
+        error.response.url === request.url
+      ) {
+        // Some backend services lose the original method and headers, so we need to do retry with cloned request
```

Comment:
Tämä request-kloonaus-viritys on tehty valintojen toteuttamisessa koska siellä otetaan tunnistautuneita yhteyksiä eri tavalla käyttäytyviin legacy-palveluihin. Tässä tapauksessa ei pitäis olla tarpeen kun tiedot tulee oletettavasti sieltä yhdestä bäkkäri-apista.